Add "Alternate" mod for osu! ruleset

LeNitrous opened the pull request:
Supersedes #9606 as an effort splitting up the taiko and osu! ruleset implementations. The taiko implementation will later be PR'ed in a separate branch.

**Considerations:**
- ~~For testing to be possible, `ModTestScene` should allow passing `ReplayFrames` for testing input-specific actions where autoplay cannot be used.~~
- Currently, the cursor flashes red when an invalid input was pressed for feedback. Maybe the keycounter as well?

Preview:

https://user-images.githubusercontent.com/20495991/151666945-aeccda60-a2d8-4b72-8463-8cc695a35c60.mp4

LeNitrous committed:
add Alternate Mod

bdach commented on osu.Game/Rulesets/Mods/ModAlternate.cs:
```diff
@@ -0,0 +1,77 @@
+    {
+        public override string Name => @"Alternate";
+        public override string Acronym => @"AL";
+        public override string Description => @"Never hit the same key twice!";
```
not sure how i missed that in the original PR, but this description is just... not right, as it can be understood as never being allowed to press a key more than once. i suggest the following clarification:

```suggestion
        public override string Description => @"Don't use the same key twice in a row!";
```

frenzibyte replied:
Might only be me, but `Never` reads better than `Don't` in this case. I think going with `Never hit the same key twice in a row!` works best.

bdach commented on osu.Game/Rulesets/Mods/ModAlternate.cs:
```diff
@@ -0,0 +1,77 @@
+        where THitObject : HitObject
+        where TAction : struct
+    {
+        public bool CanIntercept => !isBreakTime.Value;
```
wasn't in original PR, needs xmldoc or a rename (not clear what is being "intercepted")

bdach commented on osu.Game/Rulesets/Mods/ModAlternate.cs:
```diff
@@ -0,0 +1,77 @@
+
+        protected abstract bool OnPressed(TAction key);
+
+        protected abstract void OnReleased(TAction key);
```
wasn't in original PR, not sure why this one exists or what it would ever be used for

bdach commented:
in general not sure about the abstract structure. difficult to say if it will hold up on just one ruleset alone. maybe better to remove it for now and keep the entire implementation to one file for the time being.

as for testing, there isn't much more support needed than what's already there. it's already possible to play arbitrary replays on a `Player` instance via `Player.DrawableRuleset.SetReplayScore()`. hooking that up to `ModTestData` used in `ModTestScene`s is not much more work.

LeNitrous commented:
> in general not sure about the abstract structure. difficult to say if it will hold up basing on just one ruleset alone. maybe better to remove it for now and keep the entire implementation to one file for the time being.

The plan was to make a new branch off of this PR for the taiko implementation hence why the abstractions are present in `ModAlternate`.

LeNitrous committed:
remove unused method and fix description

bdach commented:
> The plan was to make a new branch off of this PR for the taiko implementation hence why the abstractions are present in `ModAlternate`.

yes, but the original taiko implementation had issues and therefore will have to be amended. and it's hard to tell if an abstract structure laid down here will still continue hold up if/when that taiko implementation lands.

put it this way: i'd prefer this to be as simple as possible for now and we can talk abstraction when an abstraction is fully evident and working.

Semro commented:
The cursor flashes red when the player already pressed the same key before hitting the first object but doesn't trigger miss. I suppose it shouldn't flash in that case. You know, some players like to smash keys before gameplay.

bdach commented:
> The cursor flashes red when the player already pressed the same key before hitting the first object but doesn't trigger miss. I suppose it shouldn't flash in that case. You know, some players like to smash keys before gameplay.

sounds like the special path for breaks already present here should also include time before first object, then.

LeNitrous committed:
move `ModAlternate` to `OsuModAlternate` and check if intro has ended

LeNitrous committed:
make xmldoc more verbose

LeNitrous committed:
allow test scenes to specify replays manually

LeNitrous committed:
add tests

bdach commented on osu.Game.Rulesets.Osu/Mods/OsuModAlternate.cs:
```diff
@@ -1,31 +1,63 @@
+        public override ModType Type => ModType.DifficultyIncrease;
+        public override IconUsage? Icon => FontAwesome.Solid.Keyboard;
+
+        /// <summary>
+        /// Whether incoming input must be checked by <see cref="InputInterceptor"/> before it is passed to gameplay.
+        /// </summary>
+        public bool CanIntercept => !isBreakTime.Value && introEnded;
```
i'm not sure this is much better/clearer than before. can we just rename this to something obvious like `PlayerMustAlternate` or something?

also, i forgot about it before, but remembered now but some maps have storyboarded outros (and lazer supports this already) - so outro start should probably get checked here, too.

LeNitrous replied:
There's no hit objects in storyboard outros so this should be sufficient.
I also do agree the xmldoc still doesn't help. I'm bad at naming and explaining things so I'll take up on your suggestion.

bdach replied:
sure, but the cursor will still flash red if you happen to spam keys in outro, would it not? unless you're saying that storyboard outro is also a break (which it may well be, i can't say i went and checked).

bdach replied:
never mind, i checked and it works as expected, because `IsBreakTime` also includes score processor completion for some reason

https://github.com/ppy/osu/blob/4132f67629467f5a5f2b2b480b391525bfe7d95e/osu.Game/Screens/Play/BreakTracker.cs#L56-L58

bdach commented on osu.Game.Rulesets.Osu/Mods/OsuModAlternate.cs:
```diff
@@ -1,31 +1,63 @@
+            drawableRuleset.KeyBindingInputManager.Add(new InputInterceptor(this));
+
+            var firstHitObject = ruleset.Objects.FirstOrDefault();
+            earliestStartTime = (firstHitObject?.StartTime ?? 0) - (firstHitObject?.HitWindows.WindowFor(HitResult.Meh) ?? 0);
```
probably fine for now, but i'm just going to note that if this gets generalised somehow then it should probably be using something like this:

https://github.com/ppy/osu/blob/4132f67629467f5a5f2b2b480b391525bfe7d95e/osu.Game/Rulesets/Scoring/HitWindows.cs#L177

LeNitrous committed:
add readonly modifier

bdach commented on osu.Game/Tests/Visual/ModTestScene.cs:
```diff
@@ -72,6 +95,12 @@ protected class ModTestData
             /// </summary>
             public bool Autoplay = true;
 
+            /// <summary>
+            /// The frames to use for replay. <see cref="Autoplay"/> must be set to false.
+            /// </summary>
+            [CanBeNull]
+            public List<ReplayFrame> Frames;
```
why `List<ReplayFrame>` instead of just a `Replay`?

LeNitrous replied:
Its much more simpler to pass the frames itself since there's no need to modify anything under `Replay.ScoreInfo`.

bdach replied:
i suppose? would probably name this `ReplayFrames` rather than `Frames` myself, though

LeNitrous committed:
rename `Frames` to `ReplayFrames`

LeNitrous committed:
rename `CanIntercept` to `ShouldAlternate`

LeNitrous committed:
remove unused using

LeNitrous commented on osu.Game.Rulesets.Osu/Mods/OsuModAlternate.cs:
```diff
@@ -0,0 +1,92 @@
+            ruleset = drawableRuleset;
+            drawableRuleset.KeyBindingInputManager.Add(new InputInterceptor(this));
+
+            var firstHitObject = ruleset.Objects.FirstOrDefault();
+            earliestStartTime = (firstHitObject?.StartTime ?? 0) - (firstHitObject?.HitWindows.WindowFor(HitResult.Meh) ?? 0);
```
Its there just to be really sure because players unconsciously spam their keys before the first hit object.

bdach committed:
Privatise `shouldAlternate`

bdach commented:
I've pushed one small change (privatising `shouldAlternate` which was needlessly public).

Diff looks palatable now but there are two soft UX decisions I'm not sure of:

* Making this mod be in the "difficulty increase" section - most lazer-exclusive mods have been landing in "fun", so doing this is rather rare (the only other case of this thus far looks to be `OsuModBlinds`, and even then it's multi-modded with flashlight)
* Currently if you miss the alternating state doesn't reset, which means you can only ever recover using the other key if you miss which can throw you off. The alternative would be to reset the alternating state on every miss, although that would allow the player to cheese half the object by not alternating. Not sure which behaviour is more correct, maybe it should be a setting. Or a follow up item.

All that is to say that I'm going to defer on the points above because I'm not sure I have a well-formed opinion on these.

mk56-spn commented:
> * Currently if you miss the alternating state doesn't reset, which means you can only ever recover using the other key if you miss which can throw you off. The alternative would be to reset the alternating state on every miss, although that would allow the player to cheese half the objects by just not alternating. Not sure which behaviour is more correct, maybe it should be a setting. Or a follow up item.
> 
> All that is to say that I'm going to defer on the points above because I'm not sure I have a well-formed opinion on these.

On those last few things i think it should go through as is and ill happily playtest it to death seeing as full alt anyways and main lazer, and then i can report back with info on whether further settings would help, idk just an idea

LeNitrous commented:
> Making this mod be in the "difficulty increase" section - most lazer-exclusive mods have been landing in "fun", so doing this is rather rare (the only other case of this thus far looks to be OsuModBlinds, and even then it's multi-modded with flashlight)

There are reasons why I came to this conclusion:
- It handicaps the player by hampering their inputs.
- It forces the player to use a different style of play.
- For visibility.

Generally speaking, it doesn't make sense for it to be in the "Fun" section. Although it can be said that other mods in said section can arguably be a "difficulty increase" as they can either hamper a player visually or audibly. However, the ways they can introduce a handicap to the player are not of the same vein as to "alternating" where in its case, it can help improve a player's skill in a practical sense.

> Currently if you miss the alternating state doesn't reset, which means you can only ever recover using the other key if you miss which can throw you off. The alternative would be to reset the alternating state on every miss, although that would allow the player to cheese half the objects by just not alternating. Not sure which behaviour is more correct, maybe it should be a setting. Or a follow up item.

There are also several things unaccounted for such as:
- Should we reset the alternating state on streams or stream-like patterns (objects whose time in-between are of 1/4ths rhythm-wise).
    - If this were the case, then how would we deal with this in a low BPM setting where the time in-between in milliseconds is long enough should making this possible may render the mod useless.
    
After dealing with the taiko version of this mod, there isn't really a well-defined standard as to what counts as "alternating" or not. I'd say we can leave it as is and receive input later on.

peppy commented:
Please put the mod in the `Conversion` type for now.

peppy commented:
> * Currently if you miss the alternating state doesn't reset, which means you can only ever recover using the other key if you miss which can throw you off. The alternative would be to reset the alternating state on every miss, although that would allow the player to cheese half the objects by just not alternating. Not sure which behaviour is more correct, maybe it should be a setting. Or a follow up item.

Not 100% on this either. My initial thought would be that it should reset the alternating state to allow either key after a certain delay, and also after a miss. Maybe fine to get it out and wait for the feedback to come in? Would be easier to add such features as a follow-up rather than expanding the scope of this PR further.

peppy committed:
Move alternate mod to "conversion" category

peppy commented on osu.Game.Rulesets.Osu/Mods/OsuModAlternate.cs:
```diff
@@ -0,0 +1,93 @@
+
+        public void Update(Playfield playfield)
+        {
+            if (!introEnded)
+                introEnded = playfield.Clock.CurrentTime > earliestStartTime;
```
I don't understand this requirement, can you explain why this is required in addition to the above `earliestStartTime` calculation?

peppy committed:
Sanitise interceptor logic to now require two separate check paths

peppy committed:
Only handle `LeftButton` and `RightButton` actions

There are definitely going to be other actions used in the future, which
would immediately cause this mod to fail. Limiting handling to
left/right buttons only is the correct way forward.

peppy commented:
I've moved the mod to the correct section and also fixed up some logic issues with it (it was handling all `OsuActions`, not just left/right button; the interceptor was doing two checks on the parent `Mod` in the same line which only made logic more confusing to follow).

I'm pretty okay with this, but holding off approving until I understand wether the special intro handling is required. I think it probably isn't and should be removed.

LeNitrous commented:
> I'm pretty okay with this, but holding off approving until I understand wether the special intro handling is required. I think it probably isn't and should be removed.

I suppose I named it where it can be confusing. The intent was to start checking for input after the "earliest hit object start time" instead of relying on the arbitrary gameplay start time as players tend to spam input before the beatmap actually begins. Think of it as a grace period.

peppy commented:
How is that different from using the judgement time allowance of the first hitobject, though?

LeNitrous commented:
I'm sorry, I can't follow what you've said on the last reply. I suppose what you mean is a better way of checking for input is checking it during the hit object's judgement?

peppy committed:
Avoid checking gameplay clock time in `Update` method

peppy commented:
Right, I understand the issue now (you were needing to access the gameplay clock's time so had it separated). I've pushed a fix for this which handles things in a slightly better way. Let me know how it looks.

LeNitrous commented:
This is definitely better. LGTM

smoogipoo commented:
Looks like this still allows sliders to be followed with the incorrect key :smile:

peppy commented:
Also noticed that, but wasn't sure it's a blocker for the mod. You still need to start with the correct key. Maybe best left for follow-up discussion?

smoogipoo commented:
I think it's okay for now, sure.

peppy commented:
Would like a second check over this before merging.
